flagext/map.go

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package flagext
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"gopkg.in/yaml.v3"
+)
+
+// LimitsMap is a flag.Value implementation that looks like a generic map, holding float64s, ints, or strings as values.
+type LimitsMap[T float64 | int | string] struct {
+	data      map[string]T
+	validator func(k string, v T) error
+}
+
+func NewLimitsMap[T float64 | int | string](validator func(k string, v T) error) LimitsMap[T] {
+	return NewLimitsMapWithData(make(map[string]T), validator)
+}
+
+func NewLimitsMapWithData[T float64 | int | string](data map[string]T, validator func(k string, v T) error) LimitsMap[T] {
+	return LimitsMap[T]{
+		data:      data,
+		validator: validator,
+	}
+}
+
+// IsInitialized returns true if the map is initialized.
+func (m LimitsMap[T]) IsInitialized() bool {
+	return m.data != nil
+}
+
+// String implements flag.Value
+func (m LimitsMap[T]) String() string {
+	out, err := json.Marshal(m.data)
+	if err != nil {
+		return fmt.Sprintf("failed to marshal: %v", err)
+	}
+	return string(out)
+}
+
+// Set implements flag.Value
+func (m LimitsMap[T]) Set(s string) error {
+	newMap := make(map[string]T)
+	if err := json.Unmarshal([]byte(s), &newMap); err != nil {
+		return err
+	}
+	return m.updateMap(newMap)
+}
+
+func (m LimitsMap[T]) Read() map[string]T {
+	return m.data
+}
+
+// Clone returns a copy of the LimitsMap.
+func (m LimitsMap[T]) Clone() LimitsMap[T] {
+	newMap := make(map[string]T, len(m.data))
+	for k, v := range m.data {
+		newMap[k] = v
+	}
+	return LimitsMap[T]{data: newMap, validator: m.validator}
+}
+
+// UnmarshalYAML implements yaml.Unmarshaler.
+func (m LimitsMap[T]) UnmarshalYAML(value *yaml.Node) error {
+	newMap := make(map[string]T)
+	if err := value.Decode(newMap); err != nil {
```

Comment:
Here's the one change; the Mimir implementation uses:
```
if err := value.DecodeWithOptions(newMap, yaml.DecodeOptions{KnownFields: true}); err != nil {
```

DecodeWithOptions is an unmerged contribution, that comes from a fork and is not part of core yaml.v3:
https://github.com/grafana/mimir/blob/9d6eb2312b070892152aab1c36566407bcb411a3/go.mod#L308-L311

`KnownFields: true` has no effect here, since we're decoding into a map. That flag only makes sense when decoding into structs. That means, we can avoid the fork and stick to only the only the core `yaml.v3` API surface, which is more portable.